File: examples/node/apps/agent/registry.ts
import {
  BaseAgenticCase,
  type AgenticContext,
} from "../../core/agentic.case";
import {
  type AgenticCatalogEntry,
  type AgenticCaseRef,
  type AgenticRegistry,
  type AppCaseSurfaces,
} from "../../core/shared/app_host_contracts";
import { createDataPackage, createJsonFileStore } from "../../packages/data/index";
import { type Task } from "../../cases/tasks/task_create/task_create.domain.case";
import { TaskCreateApi } from "../../cases/tasks/task_create/task_create.api.case";
import { TaskCreateAgentic } from "../../cases/tasks/task_create/task_create.agentic.case";
import { TaskListApi } from "../../cases/tasks/task_list/task_list.api.case";
import { TaskListAgentic } from "../../cases/tasks/task_list/task_list.agentic.case";
import { TaskMoveApi } from "../../cases/tasks/task_move/task_move.api.case";
import { TaskMoveAgentic } from "../../cases/tasks/task_move/task_move.agentic.case";
import { StreamableHttpAppMcpAdapter } from "./mcp_http";
import { StdioAppMcpAdapter } from "./mcp_stdio";

export interface AgentConfig {
  port?: number;
  dataDirectory?: string;
}

function toPublishedToolName(entry: AgenticCatalogEntry): string {
  if (entry.isMcpEnabled) {
    return entry.definition.mcp?.name ?? entry.definition.tool.name;
  }

  return entry.definition.tool.name;
}

export function createRegistry(config: AgentConfig = {}) {
  const data = createDataPackage(config.dataDirectory);
  const taskStore = createJsonFileStore<Task[]>({
    filePath: data.defaultFiles.tasks,
    fallbackData: [],
  });

  const cases = {
    tasks: {
      task_create: {
        api: TaskCreateApi,
        agentic: TaskCreateAgentic,
      },
      task_list: {
        api: TaskListApi,
        agentic: TaskListAgentic,
      },
      task_move: {
        api: TaskMoveApi,
        agentic: TaskMoveAgentic,
      },
    },
  } satisfies Record<string, Record<string, AppCaseSurfaces>>;

  const providers = {
    port: config.port ?? 3001,
    taskStore,
    mcpAdapters: {
      stdio: new StdioAppMcpAdapter(),
      http: new StreamableHttpAppMcpAdapter(),
    },
  };

  const packages = {
    data,
  };

  const registeredCases = cases as Record<string, Record<string, AppCaseSurfaces>>;

  const listAgenticCases = (): AgenticCaseRef[] =>
    Object.entries(cases).flatMap(([domain, domainCases]) =>
      Object.entries(domainCases)
        .filter(([, surfaces]) => Boolean(surfaces.agentic))
        .map(([caseName]) => ({ domain, caseName }))
    );

  const getAgenticSurface = (ref: AgenticCaseRef) =>
    registeredCases[ref.domain]?.[ref.caseName]?.agentic;

  const instantiateAgentic = (
    ref: AgenticCaseRef,
    ctx: AgenticContext
  ): BaseAgenticCase => {
    const AgenticSurface = getAgenticSurface(ref);

    if (!AgenticSurface) {
      throw new Error(
        `Agentic surface not found for ${ref.domain}/${ref.caseName}`
      );
    }

    return new AgenticSurface(ctx) as BaseAgenticCase;
  };

  const buildCatalog = (ctx: AgenticContext): AgenticCatalogEntry[] =>
    listAgenticCases().map((ref) => {
      const instance = instantiateAgentic(ref, ctx);
      const definition = instance.definition();
      const requiresConfirmation = instance.requiresConfirmation();

      const entry: AgenticCatalogEntry = {
        ref,
        publishedName: "",
        definition,
        isMcpEnabled: instance.isMcpEnabled(),
        requiresConfirmation,
        executionMode:
          definition.policy?.executionMode ??
          (requiresConfirmation ? "manual-approval" : "direct-execution"),
      };

      entry.publishedName = toPublishedToolName(entry);
      return entry;
    });

  const resolveTool = (toolName: string, ctx: AgenticContext) => {
    const normalized = toolName.trim();

    return buildCatalog(ctx).find(
      (entry) =>
        entry.publishedName === normalized ||
        entry.definition.tool.name === normalized
    );
  };

  const listMcpEnabledTools = (ctx: AgenticContext) =>
    buildCatalog(ctx).filter((entry) => entry.isMcpEnabled);

  return {
    _cases: cases,
    _providers: providers,
    _packages: packages,
    listAgenticCases,
    getAgenticSurface,
    instantiateAgentic,
    buildCatalog,
    resolveTool,
    listMcpEnabledTools,
  } satisfies AgenticRegistry;
}

export type AgentRegistry = ReturnType<typeof createRegistry>;
export type AgentCases = AgentRegistry["_cases"];
export type AgentProviders = AgentRegistry["_providers"];
export type AgentPackages = AgentRegistry["_packages"];

File: examples/kotlin/apps/agent/registry.kt
package app.protocol.examples.kotlin.apps.agent

import app.protocol.examples.kotlin.cases.tasks.task_create.TaskCreateAgentic
import app.protocol.examples.kotlin.cases.tasks.task_create.TaskCreateApi
import app.protocol.examples.kotlin.cases.tasks.task_list.TaskListAgentic
import app.protocol.examples.kotlin.cases.tasks.task_list.TaskListApi
import app.protocol.examples.kotlin.cases.tasks.task_move.TaskMoveAgentic
import app.protocol.examples.kotlin.cases.tasks.task_move.TaskMoveApi
import app.protocol.examples.kotlin.core.agentic.AgenticContext
import app.protocol.examples.kotlin.core.shared.AgenticCatalogEntry
import app.protocol.examples.kotlin.core.shared.AgenticCaseRef
import app.protocol.examples.kotlin.core.shared.AgenticRegistry
import app.protocol.examples.kotlin.core.shared.AppCaseSurfaces
import app.protocol.examples.kotlin.packages.data.JsonFileStore
import app.protocol.examples.kotlin.packages.data.createDataPackage
import app.protocol.examples.kotlin.packages.data.createJsonFileStore
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.JsonObject

data class AgentConfig(
    val port: Int = 3001,
    val dataDirectory: String? = null,
)

data class AgentRegistry(
    override val _cases: Map<String, Map<String, AppCaseSurfaces>>,
    override val _providers: Map<String, Any?>?,
    override val _packages: Map<String, Any?>?,
) : AgenticRegistry {
    override fun listAgenticCases(): List<AgenticCaseRef> =
        _cases.flatMap { (domain, domainCases) ->
            domainCases
                .filterValues { surfaces -> surfaces.agentic != null }
                .keys
                .map { caseName -> AgenticCaseRef(domain = domain, caseName = caseName) }
        }

    override fun getAgenticSurface(ref: AgenticCaseRef) = _cases[ref.domain]?.get(ref.caseName)?.agentic

    override fun instantiateAgentic(
        ref: AgenticCaseRef,
        ctx: AgenticContext,
    ) = (getAgenticSurface(ref) ?: error("Agentic surface not found for ${ref.domain}/${ref.caseName}"))(ctx)
        as app.protocol.examples.kotlin.core.agentic.BaseAgenticCase<*, *>

    override fun buildCatalog(ctx: AgenticContext): List<AgenticCatalogEntry<*, *>> =
        listAgenticCases().map { ref ->
            val instance = instantiateAgentic(ref, ctx)
            val definition = instance.definition()
            val requiresConfirmation = instance.requiresConfirmation()
            val isMcpEnabled = instance.isMcpEnabled()
            val entry =
                AgenticCatalogEntry(
                    ref = ref,
                    publishedName = "",
                    definition = definition,
                    isMcpEnabled = isMcpEnabled,
                    requiresConfirmation = requiresConfirmation,
                    executionMode = definition.policy?.executionMode
                        ?: if (requiresConfirmation) "manual-approval" else "direct-execution",
                )
            entry.copy(
                publishedName = definition.mcp?.name?.takeIf { isMcpEnabled && it.isNotBlank() } ?: definition.tool.name,
            )
        }

    override fun resolveTool(toolName: String, ctx: AgenticContext): AgenticCatalogEntry<*, *>? {
        val normalized = toolName.trim()
        return buildCatalog(ctx).firstOrNull { entry ->
            entry.publishedName == normalized || entry.definition.tool.name == normalized
        }
    }

    override fun listMcpEnabledTools(ctx: AgenticContext): List<AgenticCatalogEntry<*, *>> =
        buildCatalog(ctx).filter { it.isMcpEnabled }
}

private fun createTaskStoreProvider(
    taskStore: JsonFileStore<List<JsonObject>>,
): Map<String, Any?> {
    val read: suspend () -> List<JsonObject> = { taskStore.read() }
    val write: suspend (List<JsonObject>) -> Unit = { value -> taskStore.write(value) }
    val reset: suspend () -> Unit = { taskStore.reset() }
    val update: suspend (suspend (List<JsonObject>) -> List<JsonObject>) -> List<JsonObject> =
        { updater -> taskStore.update(updater) }

    return mapOf(
        "read" to read,
        "write" to write,
        "reset" to reset,
        "update" to update,
    )
}

fun createRegistry(config: AgentConfig = AgentConfig()): AgentRegistry {
    val data = createDataPackage(config.dataDirectory)
    val taskStore =
        createJsonFileStore(
            filePath = data.defaultFiles.tasks,
            fallbackData = emptyList<JsonObject>(),
            serializer = ListSerializer(JsonObject.serializer()),
        )

    return AgentRegistry(
        _cases = mapOf(
            "tasks" to mapOf(
                "task_create" to AppCaseSurfaces(
                    api = { ctx -> TaskCreateApi(ctx) },
                    agentic = { ctx -> TaskCreateAgentic(ctx) },
                ),
                "task_list" to AppCaseSurfaces(
                    api = { ctx -> TaskListApi(ctx) },
                    agentic = { ctx -> TaskListAgentic(ctx) },
                ),
                "task_move" to AppCaseSurfaces(
                    api = { ctx -> TaskMoveApi(ctx) },
                    agentic = { ctx -> TaskMoveAgentic(ctx) },
                ),
            ),
        ),
        _providers = mapOf(
            "port" to config.port,
            "taskStore" to createTaskStoreProvider(taskStore),
            "mcpAdapters" to mapOf(
                "stdio" to StdioAppMcpAdapter(),
                "http" to StreamableHttpAppMcpAdapter(),
            ),
        ),
        _packages = mapOf(
            "data" to data,
        ),
    )
}
